[c10d][fr] Improve FR dump robustness with all watchdog broadcast wait and more frequent store check

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150652

When debugging FR missing dump and missing dump logs, I have couple initial findings:
1. On the same rank, if a second watchdog timeout triggers on a different PG(or subPG), that watchdog thread will immediately throw exception instead of sleeping. We want to fix that by still making the watchdog thread to wait for 1 min.
2. The FR dump takes about 900ms to 1200ms so, we are not checking the store frequently enough. But instead of changing the frequency from 1sec to 300ms, we finally decided to just let all ranks just sleep for 1 min universally rather than using a promise.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[c10d][fr] Improve FR dump robustness with all watchdog broadcast wait and more frequent store check

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2141,14 +2141,11 @@ void ProcessGroupNCCL::broadcastDumpSignal() {
-  if (shouldDump_.load()) {
-    // already signaled dump, skipping signal again and wait for the dump
-    // future.
-    return;
```
Are we saying that the `return` here causes a second watchdog thread to early kill the process?

fduwjj replied:
yes

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -941,7 +941,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   waitTimeoutDumpInMilSec_ =
       getCvarInt(TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC, 60 * 1000 /*60 Sec*/);
-  coordCheckIntervalMilSec_ = getCvarInt(TORCH_NCCL_COORD_CHECK_MILSEC, 1000);
+  coordCheckIntervalMilSec_ = getCvarInt(TORCH_NCCL_COORD_CHECK_MILSEC, 300);
```
Curious: why would the checking frequency matter?

fduwjj replied:
because while one rank is sleeping, other ranks already finished the dump the FR and exit early. And the trainer just killed the training process.

fduwjj replied:
so basically I noticed that for some missing dump cases, we don't even print the log where we way that we observed the dump signals from TCPStore, while other ranks do report that, this indicates that this rank might be just sleeping? If we think this is too aggressive, I can change it to larger value?

fduwjj replied:
> Which thread finishes dumping the FR -- Heartbeat or watchdog?
Right now, Heartbeat thread.

> Which thread throws an exception which causes torchrun to broadcast killing to other ranks?
Watchdog thread

> Which thread keeps the TCPStore server alive so that other ranks can sense the dump request?
For this case, as long as rank 0 is alive, TCPStore server is alive.

I think you confused about thread coordination and rank coordination. Here I am talking about rank coordination, basically other ranks exit/throw exception early so we are seeing logs message below:

> NCCL INFO [Service thread] Connection closed by localRank 4
> NCCL INFO [Service thread] Connection closed by localRank 1

fduwjj replied:
So basically this is beyond thread level, so it does not matter which thread is doing the dump. That thread needs to get dump signal from TCPStore because the rank itself is not aware of nccl watchdog timeout.

fduwjj committed:
Update on "[c10d][fr] Improve FR dump robustness with all watchdog broadcast wait and more frequent store check"


When debugging FR missing dump and missing dump logs, I have couple initial findings:
1. On the same rank, if a second watchdog timeout triggers on a different PG(or subPG), that watchdog thread will immediately throw exception instead of sleeping. We want to fix that by still making the watchdog thread to wait for 1 min.
2. The FR dump takes about 900ms to 1200ms so, we are not checking the store frequently enough. So I change the frequency from 1sec to 300ms.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -1169,9 +1169,6 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // timeout for the dump to finish.
   int waitTimeoutDumpInMilSec_;
 
-  // promise to coordinate flight recorder dump.
-  std::promise<void> promiseFlightRecorderDump_;
-
```
LGTM.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2146,20 +2144,13 @@ void ProcessGroupNCCL::broadcastDumpSignal() {
-  // Give time for dumping before throwing exception
-  auto start = std::chrono::steady_clock::now();
-  auto status = promiseFlightRecorderDump_.get_future().wait_for(
+  // Give time for dumping before throwing exception for all ranks.
+  // It is hard to presume or control what the pattern of watchdog might look
+  // like, so it is better to let all ranks universally sleep for a short period
+  // of time, in this case, 60 seconds, which is also the maximum time we leave
+  // for FR dump.
+  std::this_thread::sleep_for(
       std::chrono::milliseconds(waitTimeoutDumpInMilSec_));
```
You may want to put the sleep somewhere else than inside the `broadcastDumpSignal` function, so that others calling this function would be sure what it does. Like, one-level up maybe?

kwen2501 replied:
btw, do we need "all ranks" to sleep? Can it be just the rank that's throwing exception?

fduwjj replied:
sounds good to me.

kwen2501 reviewed (approved):
LGTM. I left one comment.

fduwjj committed:
Update on "[c10d][fr] Improve FR dump robustness with all watchdog broadcast wait and more frequent store check"


When debugging FR missing dump and missing dump logs, I have couple initial findings:
1. On the same rank, if a second watchdog timeout triggers on a different PG(or subPG), that watchdog thread will immediately throw exception instead of sleeping. We want to fix that by still making the watchdog thread to wait for 1 min.
2. The FR dump takes about 900ms to 1200ms so, we are not checking the store frequently enough. But instead of changing the frequency from 1sec to 300ms, we finally decided to just let all ranks just sleep for 1 min universally rather than using a promise.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot -i merge
